### DISCUSSION TITLE
Problem with federation

### DISCUSSION THREAD

**Discussion by ramiel** [MichalLytek/type-graphql#1376] [Q&A] [answered]
**Describe the Bug**
I have a problem using type-graphql and federation (with any federation server). I created a small reproduction repository. Basically I think the main problem is that `__resolveReference` function is never called even if the request is correctly routed to the correct subgraph

**To Reproduce**
Clone and run https://github.com/ramiel/typegrapqhl-federation-issue
Read the readme file for a detailed explanation.

**Expected Behavior**
Code should work as the example provided by you

**Environment (please complete the following information):**

- OS: MacOS
- Node 16
- Package version 2.0.0-beta
- TypeScript version 4.8

**Comment by ramiel** [comment]
I find out the problem is actually present also in your `example/apollo-federation` code. To show it you can run the query

```graphql
query Me {
  me {
    reviews {
      product {
        name
      }
    }
  }
}
```

This shows the federation is actually not working. @MichalLytek can we move this to issues instead of a discussion?

**Comment by ramiel** [comment]
The error is present with the example query as well: https://github.com/MichalLytek/type-graphql/blob/new-major-release/examples/apollo-federation/examples.graphql

**Comment by MichalLytek** [comment, marked as answer]
@ramiel `__resolveReference` is not a part for TypeGraphQL. It's defined by user and it's his own responsibility to build the apollo schema.

The problem is that with old apollo federation and `buildApolloFederationSchema`, we could add reference resolvers after building the schema, and it was working well. Now with `buildSubgraphSchema` it expects reference resolvers to be already provided with the schema.
So we have to merge TypeGraphQL resolvers with reference resolvers somehow. My approach is like this:
```ts
import deepMerge from "lodash.merge";

const federatedSchema = buildSubgraphSchema({
  typeDefs: gql(printSchemaWithDirectives(schema)),
  resolvers: deepMerge(createResolversMap(schema) as any, referenceResolvers),
});
```
I will update the example with this, thanks for noticing the issue.

**Comment by ramiel** [comment]
Thank you for the reply. I'm going to try this. Understanding all the differences between federation v1 and v2 is a pain 😄
